### PR TITLE
Feature/149 trips and expenses

### PR DESCRIPTION
**Pull request by tbjoern**
Beinhaltet branches refactor/150_expenses und feature/150_trip_details
User stories sind dementsprechend #149 und #150

### CODE REVIEW

File: app/controllers/expenses_controller.rb [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+
+  def edit
+    if @expense.status == 'applied'
+      redirect_to trip_path(@expense.trip)
+      flash[:error] = I18n.t('expense.applied')
+    else
+    end
+  end
+
+  def create
```
Comment (swt-r0ckz) [inline app/controllers/expenses_controller.rb]: im else case wird hier nichts ausgeführt

File: app/models/expense.rb [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+  enum status: [:saved, :applied, :accepted, :declined]
+
+  def first_name
+    user.first_name
+  end
+
+  def last_name
+    user.last_name
+  end
+
```
Comment (swt-r0ckz) [inline app/models/expense.rb]: brauchst du wirklich first- und last name einzeln oder ginge auch 'user.name' ?

File: app/models/expense.rb [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+  def get_signature
+    user.signature
+  end
+
```
Comment (swt-r0ckz) [inline app/models/expense.rb]: was macht diese methode hier?

File: app/models/expense.rb [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+    if time_start && !time_start.match(/\A(?:[0-1]?[0-9]|2[0-3]):[0-5][0-9]/)
+      errors.add(:time_start, I18n.t('activerecord.errors.models.expense.attributes.time.format'))
+    end
+    if time_end && !time_end.match(/\A(?:[0-1]?[0-9]|2[0-3]):[0-5][0-9]/)
+      errors.add(:time_end, I18n.t('activerecord.errors.models.expense.attributes.time.format'))
+    end
+  end
+
+  
+end
```
Comment (swt-r0ckz) [inline app/models/expense.rb]: Das letzte end fühlt sich hier einsam

File: app/views/expenses/_expense.html.erb [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+
+<p>
+  <strong><%= t('.general_advance') %>:</strong>
+  <%= expense.general_advance %>
+</p>
+
+<p>
+  <strong><%= t('.signature') %>:</strong>
+  <%= t(expense.signature.to_s) %>
+</p>
```
Comment (swt-r0ckz) [inline app/views/expenses/_expense.html.erb]: Das sieht komisch aus, wenn Unterschreiben: true (oder false) angezeigt wird. Könntest du das noch durch ja/nein ersetzen?

File: config/locales/de.yml [hunk 6]
```diff
@@ -95,19 +95,18 @@ de:
...
         inland: "Inlandsreise"
         country: "Land"
         location_from: "Startort"
         location_via: "Zwischenhalt"
         location_to: "Zielort"
         reason: "Grund"
-        date_start: "Startet am"
-        date_end: "Endet am"
+        time_start: "Abfahrtszeit"
+        time_end: "Ankuftszeit"
```
Comment (swt-r0ckz) [inline config/locales/de.yml]: Hier fehlt ein n Anku N ftszeit

File: config/locales/de.yml [hunk 7]
```diff
@@ -164,7 +163,7 @@ de:
         workload: "Arbeitszeit"
     errors:
       models:
-        travel_expense_report:
+        expense:
           attributes:
```
Comment (swt-r0ckz) [inline config/locales/de.yml]: Die Translations der errors für die Trips fehlen noch ganz.

File: db/migrate/20160126160730_drop_table_trip_datespan.rb [hunk 8]
```diff
@@ -0,0 +1,4 @@
+class DropTableTripDatespan < ActiveRecord::Migration
+  def change
+  end
+end
```
Comment (swt-r0ckz) [inline db/migrate/20160126160730_drop_table_trip_datespan.rb]: mmhh eine leere Migration

File: spec/controllers/expenses_controller_spec.rb [hunk 9]
```diff
@@ -0,0 +1,239 @@
...
+# It only uses APIs available in rails and/or rspec-rails.  There are a number
+# of tools you can use to make these specs even more expressive, but we're
+# sticking to rails and rspec-rails APIs to keep things simple and stable.
+#
+# Compared to earlier versions of this generator, there is very limited use of
+# stubs and message expectations in this spec.  Stubs are only used when there
+# is no simpler way to get a handle on the object needed for the example.
+# Message expectations are only used when there is no simpler way to specify
+# that an instance is receiving a specific message.
+
```
Comment (swt-r0ckz) [inline spec/controllers/expenses_controller_spec.rb]: Könntest du den Kommentar bitte noch löschen.

File: app/models/expense.rb (hunk 2 above)
Comment (tbjoern) [inline app/models/expense.rb]: Die werden fuer die PDF gebraucht.

File: app/controllers/expenses_controller.rb [hunk 10]
```diff
@@ -0,0 +1,107 @@
...
+
+  def index
+    @expenses = Expense.where(user: current_user)
+  end
+
+  def show
+  end
+
+  def new
+    @trip = Trip.find(params[:trip_id])
```
Comment (Tymolc) [inline app/controllers/expenses_controller.rb]: Du machst ja als before_action :set_trip. Das macht das gleiche wie in dieser Zeile hier. Die brauchst du also nicht.

File: app/controllers/trips_controller.rb [hunk 11]
```diff
@@ -29,7 +28,6 @@ def edit
       redirect_to @trip
       flash[:error] = I18n.t('trip.applied')
     else
```
Comment (Tymolc) [inline app/controllers/trips_controller.rb]: Hier ist der else case jetzt auch leer

File: app/models/expense.rb [hunk 12]
```diff
@@ -0,0 +1,90 @@
...
+        day.dinner = false
+        day.save
+      end
+    end
+  end
+
+  private
+
+  def time_format
+    if time_start && !time_start.match(/\A(?:[0-1]?[0-9]|2[0-3]):[0-5][0-9]/)
```
Comment (Tymolc) [inline app/models/expense.rb]: Könntest du das vielleicht in eine eigene Funktion auslagern? Das is so doch etwas kompliziert zu lesen.

File: app/models/trip.rb [hunk 13]
```diff
@@ -11,17 +11,21 @@
...
-  accepts_nested_attributes_for :trip_datespans, reject_if: lambda {|attributes| attributes['days_abroad'].blank?}
-  validates :destination, presence: true
-  validates :user, presence: true
+  has_one :expense
+  validates_presence_of :destination, :user,:date_start,:date_end, :days_abroad
+  validates :days_abroad, numericality: {greater_than_or_equal_to: 0}
+  validate :start_before_end_date, :days_abroad_leq_to_total_days
   belongs_to :person_in_power, class_name: 'User'
-  has_many :travel_expense_reports
+
```
Comment (Tymolc) [inline app/models/trip.rb]: Hier könnte eine blank line raus :)

File: app/models/trip.rb [hunk 14]
```diff
@@ -29,10 +33,39 @@ class Trip < ActiveRecord::Base
...
+  def total_days
+    (date_end - date_start).to_i + 1
+  end
+
+  private
+
+  def days_abroad_leq_to_total_days
+    #TODO: Geht das besser?
+    if date_end && date_start && days_abroad && days_abroad > total_days
```
Comment (Tymolc) [inline app/models/trip.rb]: Das hab ich mich auch schon mal gefragt, aber nichts gefunden, was wirklich schöner war. Ich wäre auch interessiert, ob da jemand eine bessere Lösung hat.

File: app/views/expense_items/_show.html.erb [hunk 15]
```diff
@@ -0,0 +1 @@
+<li>Date: <%= expense_item.date.strftime("%d.%m.%Y") %> F: <%= expense_item.breakfast %>M: <%= expense_item.lunch %>A: <%= expense_item.dinner %>Description:<%= expense_item.annotation %></li>
```
Comment (Tymolc) [inline app/views/expense_items/_show.html.erb]: Gibt es einen Grund warum das hier alles in einer Zeile ist und das Datum immer auf das deutsche Format gestellt wird?

File: app/views/expenses/_expense.html.erb [hunk 16]
```diff
@@ -0,0 +1,73 @@
...
+<ul>
+  <%= render expense.expense_items %>
+</ul>
```
Comment (Tymolc) [inline app/views/expenses/_expense.html.erb]: Eine newline wäre hier schön :)

File: app/views/expenses/_form.html.erb [hunk 17]
```diff
@@ -0,0 +1,98 @@
...
+        <th><%= t('activerecord.attributes.expense_item.date') %></th>
+        <th><%= t('activerecord.attributes.expense_item.breakfast') %></th>
+        <th><%= t('activerecord.attributes.expense_item.lunch') %></th>
+        <th><%= t('activerecord.attributes.expense_item.dinner') %></th>
+        <th><%= t('activerecord.attributes.expense_item.annotation') %></th>
+      </tr>
+      <% index = 0 %>
+      <%= f.fields_for :expense_items do |t| %>
+        <tr>
+          <td><%= @expense.expense_items[index].date %></td>
```
Comment (Tymolc) [inline app/views/expenses/_form.html.erb]: Wenn das hier ein Datum ist, dann solltest du da noch ein l vorschreiben.

File: app/models/trip.rb (hunk 14 above)
Comment (tbjoern) [inline app/models/trip.rb]: +1 darauf, ich bin ewig dran verzweifelt eine gut loesung zu finden

File: app/views/trips/_form.html.erb [hunk 18]
```diff
@@ -23,25 +22,23 @@
...
+      <th><%= model_class.human_attribute_name(:date_start) %></th>
+      <th><%= model_class.human_attribute_name(:date_end) %></th>
+      <th><%= model_class.human_attribute_name(:days_abroad) %></th>
     </tr>
-    <%= f.fields_for :trip_datespans do |d| %>
     <tr>
       <td>
         <div class="control-group">
-        <%= d.text_field :start_date, required: false, class: 'form-control', 'data-provide': "datepicker", 'data-date-orientation': "bottom auto", 'data-date-format': "dd-mm-yyyy", 'data-date-autoclose': true, 'data-date-language': I18n.locale %>
+        <%= f.text_field :date_start, required: false, class: 'form-control', 'data-provide': "datepicker", 'data-date-orientation': "bottom auto", 'data-date-format': "yyyy-mm-dd", 'data-date-autoclose': true, 'data-date-language': I18n.locale %>
```
Comment (Tymolc) [inline app/views/trips/_form.html.erb]: Könntest du bei dem Datepicker noch mal schauen, wie ich das mit der Lokalisierung bei den Holidays gemacht habe? So würde er ja immer das Format yyyy-mm-dd haben.

File: app/views/trips/index.html.erb [hunk 19]
```diff
@@ -26,6 +29,9 @@
         <td><%= trip.reason %></td>
         <td><%= trip.annotation %></td>
         <td><%= trip.status %></td>
+        <td><%= trip.date_start %></td>
+        <td><%= trip.date_end %></td>
```
Comment (Tymolc) [inline app/views/trips/index.html.erb]: Hier kannst du die Daten auch noch lokalisieren.

File: config/locales/en.bootstrap.yml [hunk 20]
```diff
@@ -215,15 +215,16 @@ en:
...
+    expense:
       inland: "Inland Travel"
       country: "Country"
       location_from: "Starting in"
       location_via: "Via"
       location_to: "To"
       reason: "Reason"
-      date_start: "Start at"
+      time_start: "Depart by"
+      time_end: "Arrive by"
```
Comment (Tymolc) [inline config/locales/en.bootstrap.yml]: Wäre hier sowas wie "Departure time" und "Arrival time" vielleicht eine bessere Übersetzung?

File: spec/controllers/trips_controller_spec.rb [hunk 21]
```diff
@@ -31,19 +31,24 @@
...
   let(:valid_attributes) {
     {destination: 'NYC Conference',
-     reason: 'Hana Things',
-     annotation: 'HANA pls',
-     signature: true,
-     user: @user}
+    reason: 'Hana Things',
+    annotation: 'HANA pls',
+    date_start: Date.today,
+    date_end: Date.today+2,
```
Comment (Tymolc) [inline spec/controllers/trips_controller_spec.rb]: Ich hab letztens mal mit jemandem vom SWA-Lehrstuhl geredet und das Ergebnis war, dass es wohl besser ist in Tests statische Daten zu benutzen und lieber Corner Cases (Leap Year, Jahreswechsel ...) abzufangen. Wenn ihr damit nichts rechnet, dann sollte das aber relativ egal sein, glaube ich.

File: app/controllers/expenses_controller.rb [hunk 22]
```diff
@@ -0,0 +1,99 @@
...
+    if @expense.status == 'applied'
+      redirect_to trip_path(@expense.trip)
+      flash[:error] = I18n.t('expense.applied')
+    else
+      @expense.destroy
+      redirect_to trip_url(trip)
+      flash[:success] = I18n.t('expense.destroyed')
+    end
+  end
+
```
Comment (maikz) [inline app/controllers/expenses_controller.rb]: Best practise bei der Reihenfolge der actions wäre new, create, edit, update, destroy, hand_in

File: config/locales/de.yml (hunk 7 above)
Comment (tbjoern) [inline config/locales/de.yml]: TODO: neues Issue

File: spec/controllers/trips_controller_spec.rb (hunk 21 above)
Comment (tbjoern) [inline spec/controllers/trips_controller_spec.rb]: Es wird mit `@trip.totaldays` die Anzahl Tage berechnet, ansonsten nichts. Sollte ich das trotzdem aendern ? Was wuerde man eintragen ?

File: app/controllers/expenses_controller.rb (hunk 22 above)
Comment (tbjoern) [inline app/controllers/expenses_controller.rb]: Ist implementiert

File: spec/controllers/trips_controller_spec.rb (hunk 21 above)
Comment (Tymolc) [inline spec/controllers/trips_controller_spec.rb]: Ich glaube schon. Vielleicht sowas wie Date.new(2016, 2, 28) und Date.new(2016, 3, 1) und dann erwartest du, dass 3 rauskommt, weil dieses Jahr Schaltjahr ist.
